lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj

```diff
@@ -1617,7 +1622,14 @@ private JsonNode Json() :
 { String s; Token t; JsonNode n; }
 { 
 ( s = JsonString() { n = new TextNode(s); }
-| (t=<INTEGER_LITERAL> { n = new LongNode(Long.parseLong(t.image)); })
+| (t=<INTEGER_LITERAL> {
+  long longValue = Long.parseLong(t.image);
+  int intValue = (int)longValue;
+  if (intValue == longValue)
+    n = new IntNode(intValue);
+  else
+    n = new LongNode(longValue);
+})
```

Comment:
This bit here creates a JSON node with an Integer in it instead of a Long if the number is small enough.
This is needed because the logical type `decimal` will fail to validate if the scale or precision are Long values (even if small enough).